nx/lib/nx/random.ex

```diff
@@ -0,0 +1,390 @@
+    |> Nx.as_type({:u, 32})
+  end
+
+  @doc """
+  Splits a PRNG key into `num` new keys by adding a leading axis.
+  
+    ## Examples
+  
+      iex> key = Nx.Random.key(1701)
+      iex> Nx.Random.split(key, 2)
+      #Nx.Tensor<
+        u32[2][2]
+        [
+          [56197195, 1801093307],
+          [961309823, 1704866707]
+        ]
+      >
+  
+      iex> key = Nx.Random.key(999999999999)
+      iex> Nx.Random.split(key, 4)
+      #Nx.Tensor<
+        u32[4][2]
+        [
+          [3959978897, 4079927650],
+          [3769699049, 3585271160],
+          [3182829676, 333122445],
+          [3185556048, 1258545461]
+        ]
+      >
+  """
+  def split(key, num \\ 2) when num > 1 do
+    assert_key(key)
+
+    impl(key, Nx.iota({num, 2}))
+  end
+
+  def fold_in(key, data) when is_integer(data) do
+    assert_key(key)
+
+    impl(key, key(data))
+  end
+
+  def random_bits(key, shape \\ {1}) do
+    assert_key(key)
+
+    impl(key, Nx.iota(shape))
+  end
+
+  defnp impl(key, count) do
+    shape = Nx.shape(count)
+
+    reshaped_key = Nx.reshape(key, {2, 1})
+
+    reshaped_count =
+      Nx.reshape(count, {:auto})
+      |> Nx.as_type({:u, 32})
+
+    threefry2x32(reshaped_key, reshaped_count)
+    |> Nx.reshape(shape)
+    |> Nx.as_type({:u, 32})
+  end
+
+  # Check count
+  defn threefry2x32(key, count) do
+    transform(count, &threefry2x32_in_transform/1)
+    |> Nx.reshape({2, :auto})
+    |> Nx.as_type({:u, 32})
+    |> threefry2x32_20(key)
+    |> transform(&threefry2x32_out_transform(&1, count))
+  end
+
+  defp threefry2x32_in_transform(count) do
+    count_size = Nx.axis_size(count, 0)
+    even? = rem(count_size, 2) == 0
+
+    if even? do
+      count
+    else
+      Nx.concatenate([count, Nx.tensor([0])])
+    end
+  end
+
+  defp threefry2x32_out_transform(count, original) do
+    count_size = Nx.axis_size(original, 0)
+    even? = rem(count_size, 2) == 0
+
+    if even? do
+      count
+    else
+      count
+      |> Nx.flatten()
+      |> Nx.slice_along_axis(0, count_size, axis: 0)
+    end
+  end
+
+  defnp threefry2x32_20(xs, ks) do
+    rotations = Nx.tensor([[13, 15, 26, 6], [17, 29, 16, 24]], type: {:u, 8})
+    key1 = ks[0]
+    key2 = ks[1]
+
+    xs = Nx.add(ks, xs)
+
+    ks =
+      Nx.stack([
+        key2,
+        Nx.bitwise_xor(key1, key2)
+        |> Nx.bitwise_xor(0x1BD11BDA),
+        key1
+      ])
+      |> Nx.as_type({:u, 32})
+
+    state = {xs, ks, rotations}
+
+    {_, {nxs, _, _}} =
+      while {x = 0, state}, Nx.less(x, 5) do
+        {x + 1, rolled_loop_step(x, state)}
+      end
+
+    nxs
+  end
+
+  defnp apply_round(xs, rot) do
+    y1 = Nx.add(xs[0], xs[1])
+
+    y2 =
+      rotate_left(xs[1], rot)
+      |> Nx.bitwise_xor(y1)
+
+    Nx.stack([y1, y2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  defnp rolled_loop_step(i, {xs, ks, rs}) do
+    {k1, k2, k3} = {ks[0], ks[1], ks[2]}
+    {r1, r2} = {rs[0], rs[1]}
+
+    {_, xs, _} =
+      while {x = 0, xs, rs}, Nx.less(x, 4) do

```

Comment:
```suggestion
      while {x = 0, xs, rs}, x < 4 do
```

This works exactly the same :)